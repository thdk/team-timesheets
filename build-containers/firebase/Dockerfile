FROM node:16 AS node_base

RUN echo "NODE Version:" && node --version
RUN echo "NPM Version:" && npm --version

FROM openjdk:8-alpine
COPY --from=node_base . .

# install Firebase CLI
RUN npm install -g firebase-tools
RUN firebase setup:emulators:firestore

ENTRYPOINT ["/usr/local/bin/firebase"]