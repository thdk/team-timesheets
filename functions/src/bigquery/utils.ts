// Imports the Google Cloud client library
import { BigQuery } from '@google-cloud/bigquery';
import { BigQueryTableSchemes } from './config';

export type BigQueryField = { name: string, type: string, mode?: string };
export type BigQueryConfig = {
    dataSetId: string;
    tableId: string;
    tableIdPrefix: string,
    schemes: BigQueryTableSchemes,
    timePartitioningField?: string,
};

function validateTableSchemaAsync(table: any, schema: BigQueryField[]) {
    return table.getMetadata().then(([metadata]) => {
        if (metadata.schema) {
            console.log(`Current schema of table ${table.id} is ${JSON.stringify(metadata.schema.fields)}`);
        } else {
            console.log("Table has no schema yet.");
        }

        console.log(`Required schema of table ${table.id} is ${JSON.stringify(schema)}`);

        if (!isEqualSchema(metadata.schema.fields, schema)) {
            console.log(`Updating schema of table ${table.id}...`);
            return table.setMetadata(Object.assign(metadata, { schema }))
                .then(() => {
                    console.log("Schema updated.");
                    return table;
                }, () => {
                    console.log("Couldn't update schema of table: " + table.id);
                });
        }
        else {
            console.log(`Schema of table ${table.id}... is already up to date.`);
            return new Promise(resolve => resolve(table));
        }
    });
}

function isEqualSchema(schema1: BigQueryField[], schema2: BigQueryField[]) {
    console.log(`Comparing fields length... ${schema1.length} <> ${schema2.length}`);
    if (schema1.length !== schema2.length) return false;

    return !schema1.some((field1, index) => {

        const field2 = schema2[index];
        console.log(`Comparing field...`);
        console.log(`${JSON.stringify(field1)}`);
        console.log(`${JSON.stringify(field2)}`);

        return field1.name !== field2.name ||
            field1.type !== field2.type ||
            field1.mode !== field2.mode;
    });
}

type InsertOptions = {
    schema: BigQueryField[];
    location: string;
    timePartitioning?: {
        type: 'DAY';
        field: string;
    };
}
export function insertRowsAsync<T>(options: BigQueryConfig, rows: ReadonlyArray<T>, bigqueryClient?: BigQuery): Promise<void> {
    const { dataSetId, tableId, tableIdPrefix = "", schemes, timePartitioningField } = options;

    console.log(`Inserting ${rows.length} rows into ${tableId}`);

    let insertOptions: InsertOptions = {
        schema: schemes[tableId],
        location: 'US'
    };

    if (timePartitioningField) {
        insertOptions = {
            ...insertOptions,
            timePartitioning: {
                type: 'DAY',
                field: timePartitioningField,
            },
        }
    }

    if (!rows.length) return new Promise(resolve => resolve());

    const bigquery = bigqueryClient || new BigQuery();

    // TODO: create dataset only once since insertRows can be called multiple times simultaneously!
    return bigquery
        .dataset(dataSetId)
        .get({ autoCreate: true })
        .then(([dataset]) => {
            console.log("dataset created");
            return dataset.table(`${tableIdPrefix}${tableId}`)
                .get({ autoCreate: true, ...insertOptions })
                .then(([table]) => validateTableSchemaAsync(table, insertOptions.schema))
                .then(table => {
                    const first10kRows = rows.slice(0, 10000);
                    return table.insert(first10kRows)
                        .then(() => {
                            console.log(`Inserted ${first10kRows.length} rows of total ${rows.length} into ${tableId}`);
                            if (rows.length > 10000) {
                                console.log("Inserting next batch...")
                                return insertRowsAsync(options, rows.slice(10000));
                            }

                            console.log(`Finished inserting into ${tableId}`);
                            return true;
                        }, (error: { errors: any, name: string, response: any, message: string }) => {
                            error.errors && error.errors.forEach(e => {
                                console.log({
                                    error: error.name,
                                    message: error.message,
                                    errors: e.errors,
                                    row: e.row
                                });
                            });
                        }, (e) => {
                            console.log(e);
                            console.log("Table could not be created");
                        })
                })
        }, e => {
            console.log(e);
            console.log("Dataset could not be created");
        });

}