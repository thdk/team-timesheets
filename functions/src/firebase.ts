import * as admin from "firebase-admin";

admin.initializeApp({ credential: admin.credential.applicationDefault() });

export const db = admin.firestore();
