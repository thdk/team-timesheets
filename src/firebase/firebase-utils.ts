import { Auth, User } from "firebase/auth";

/**
 * Resolves with firbase.User if user is logged in
 * Rejects if no user is logged in
 */
export const getLoggedInUserAsync = (auth: Auth) => {
    return new Promise<User>((resolve, reject) => {
        const unsubscribe = auth.onAuthStateChanged(user => {
            unsubscribe();
            if (user) resolve(user);
            else reject("Not authenticated");
        });
    });
}