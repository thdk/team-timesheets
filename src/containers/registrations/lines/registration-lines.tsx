import * as React from "react";
import { observer } from 'mobx-react-lite';
import { Doc } from "firestorable";
import { IRegistration, IRegistrationData } from '../../../../common';
import { RegistrationLine } from "../line";
import { useProjectStore } from "../../../stores/project-store";
import { useTasks } from "../../../contexts/task-context";
import { useClients } from "../../../contexts/client-context";

export interface IRegistrationLinesProps extends React.HTMLProps<HTMLElement> {
    readonly registrations: Doc<Omit<IRegistration, "date" | "isPersisted">, Omit<IRegistrationData, "date">>[];
    readonly registrationToggleSelect?: (id: string) => void;
    readonly registrationClick?: (id: string) => void;
}

export const RegistrationLines = observer((props: IRegistrationLinesProps) => {
    const projects = useProjectStore();
    const tasks = useTasks();
    const clients = useClients();

    const { registrations, registrationToggleSelect, registrationClick } = props;

    const rows = registrations.map(r => {
        if (!r.data) return;

        const { id, data: { description = "...", project, time, task, client } } = r;

        const projectData = project ? projects.collection.get(project) : null;
        const { data: { name: projectName = "" } = {} } = projectData || {};

        const taskData = task ? tasks.get(task) : null;
        const { data: { icon = undefined, name: taskName = "" } = {} } = taskData || {};

        const clientData = client ? clients.get(client) : null;
        const { data: { name: clientName = undefined } = {} } = clientData || {};

        const line1 = projectName;
        const line2 = `${clientName ? clientName + " - " : ""}${description}`;

        return (
            <RegistrationLine
                readOnly={!registrationClick}
                icon={icon ? {
                    icon,
                    title: taskName,
                } : undefined}
                id={id}
                key={id}
                line1={line1}
                line2={line2}
                time={time}
                onClick={
                    registrationClick
                        ? () => registrationClick(id)
                        : undefined
                }
                onSelect={registrationToggleSelect
                    ? () => registrationToggleSelect(id)
                    : undefined}>
            </RegistrationLine>
        );
    });

    return <>{rows}</>;
});