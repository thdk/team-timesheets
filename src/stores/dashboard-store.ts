import { observable, reaction, computed, action, makeObservable } from "mobx";
import { ICollection, Collection, RealtimeMode } from "firestorable";

import moment from 'moment';

import { IRootStore } from "./root-store";
import { IGroupedRegistrations } from "./registration-store/registration-store";
import * as deserializer from '../../common/serialization/deserializer';
import * as serializer from '../../common/serialization/serializer';
import { TimePeriod } from "../components/time-period-select";
import { IRegistration, IRegistrationData } from "../../common/dist";
import { Firestore, query, Query, where } from "firebase/firestore";

export interface IDashboardStore {
    readonly setProjectFilter: (projectId: string | undefined) => void;
    readonly setUserFilter: (userId: string | undefined) => void;
    readonly setTaskFilter: (taskId: string | undefined) => void;
    readonly setTimePeriodFilter: (timePeriod: TimePeriod) => void;

    readonly registrationsGroupedByUser: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByTask: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByProject: IGroupedRegistrations<string>[];

    readonly projectFilterValue?: string;
    readonly taskFilterValue?: string;
    readonly userFilterValue?: string;
    readonly timePeriodFilterValue?: TimePeriod;

}

export class DashboardStore implements IDashboardStore {
    private readonly registrationsField: ICollection<IRegistration, IRegistrationData>;

    private userFilterValueField: string | undefined = undefined;
    private taskFilterValueField: string | undefined = undefined;
    private projectFilterValueField: string | undefined = undefined;
    private timePeriodFilterField: TimePeriod | undefined = undefined;

    constructor(
        rootStore: IRootStore,
        {
            firestore,
        }: {
            firestore: Firestore,
        },
    ) {
        makeObservable<DashboardStore, "userFilterValueField" | "taskFilterValueField" | "projectFilterValueField" | "timePeriodFilterField" | "startDateFilter" | "endDateFilter" | "userIdFilter" | "projectIdFilter" | "startDate" | "endDate" | "registrations">(this, {
            userFilterValueField: observable.ref,
            taskFilterValueField: observable.ref,
            projectFilterValueField: observable.ref,
            timePeriodFilterField: observable.ref,
            startDateFilter: computed,
            endDateFilter: computed,
            userIdFilter: computed,
            projectIdFilter: computed,
            registrationsGroupedByUser: computed,
            registrationsGroupedByProject: computed,
            registrationsGroupedByTask: computed,
            startDate: computed,
            endDate: computed,
            registrations: computed,
            setProjectFilter: action,
            setTaskFilter: action,
            setUserFilter: action,
            setTimePeriodFilter: action
        });

        this.registrationsField = new Collection<IRegistration, IRegistrationData>(
            firestore,
            "registrations",
            {
                realtimeMode: RealtimeMode.on,
                deserialize: deserializer.convertRegistration,
                serialize: serializer.convertRegistration,
                name: "Dashboard registrations",
            },
            {
                // logger: console.log
            },
        );

        const updateRegistrationQuery = () => {
            if (!rootStore.user.divisionUser || !this.timePeriodFilterField) {
                this.registrationsField.query = null;
            } else {
                this.registrationsField.query = ref =>
                    this.userIdFilter(
                        this.projectIdFilter(
                            this.endDateFilter(
                                this.startDateFilter(
                                    query(ref, where("deleted", "==", false))
                                )
                            )
                        )
                    ) as Query<IRegistrationData>;
            }
        };

        // update the query of the registration collection each time...
        // -- a filter changes (userId, startDate, endDate)
        // TODO: add a Filter button in UI to group all filter changes into one new search query
        reaction(() => this.timePeriodFilterField, updateRegistrationQuery);
        reaction(() => this.userFilterValue, updateRegistrationQuery);
        reaction(() => this.projectFilterValue, updateRegistrationQuery);
        reaction(() => rootStore.user.divisionUser, updateRegistrationQuery);
    }

    private get startDateFilter() {
        const startDate = this.startDate;
        return (q: Query) => startDate ? query(q, where("date", ">=", startDate)) : q;
    }

    private get endDateFilter() {
        const endDate = this.endDate;
        return (q: Query) => endDate ? query(q, where("date", "<=", endDate)) : q;
    }

    private get userIdFilter() {
        const userId = this.userFilterValue;
        return (q: Query) => userId ? query(q, where("userId", "==", userId)) : q;
    }

    private get projectIdFilter() {
        return (q: Query) => this.projectFilterValue
            ? query(q, where("project", "==", this.projectFilterValue))
            : q;
    }

    get registrationsGroupedByUser() {
        // no need to group when userId is included in the filter
        if (this.registrations.length === 0) return [];

        const registrationsPerUser = this.createRegistrationsGroup(reg => reg.userId);

        return registrationsPerUser;
    }

    get registrationsGroupedByProject() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.project || "")
            .sort((a, b) => b.totalTime - a.totalTime);;
    }

    get registrationsGroupedByTask() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.task || "")
            .sort((a, b) => b.totalTime - a.totalTime);
    }

    private get startDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().startOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().startOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').startOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().startOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').startOf('year').toDate();
            default:
                return undefined;
        }
    }

    private get endDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().endOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().endOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').endOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().endOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').endOf('year').toDate();
            default:
                return undefined;
        }
    }

    public get taskFilterValue() {
        return this.taskFilterValueField;
    }

    public get userFilterValue() {
        return this.userFilterValueField;
    }

    public get projectFilterValue() {
        return this.projectFilterValueField;
    }

    public get timePeriodFilterValue() {
        return this.timePeriodFilterField;
    }

    private get registrations() {
        return this.registrationsField.docs
            .filter(doc => doc.data!.isPersisted) // don't display drafts (TODO: move to collection.ts?)
    }

    public setProjectFilter(projectId: string | undefined) {
        this.projectFilterValueField = projectId;
    }

    public setTaskFilter(taskId: string | undefined) {
        this.taskFilterValueField = taskId;
    }

    public setUserFilter(userId: string | undefined) {
        this.userFilterValueField = userId;
    }

    setTimePeriodFilter(timePeriod: TimePeriod) {
        this.timePeriodFilterField = timePeriod;
    }

    private createRegistrationsGroup = <T>(keySelector: (reg: IRegistration) => T): IGroupedRegistrations<T>[] => {
        const groupedMap = new Map<T, IGroupedRegistrations<T>>();

        this.registrations
            .forEach(c => {
                const key = keySelector(c.data!);
                const group = groupedMap.get(key);
                if (group) {
                    group.totalTime += c.data!.time || 0;
                    group.registrations.push(c);
                }
                else {
                    groupedMap.set(key, {
                        groupKey: keySelector(c.data!),
                        registrations: [c],
                        totalTime: c.data!.time || 0,
                        isCollapsed: true
                    });
                }
            });

        return Array.from(groupedMap.values());
    }
}