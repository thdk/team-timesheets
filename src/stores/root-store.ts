import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { IProjectStore, ProjectStore } from "./project-store";
import { FavoriteStore } from "./favorite-store";

export interface IRootStore {
    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly favorites: FavoriteStore;
    readonly dashboard: IDashboardStore;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;
    public readonly favorites: FavoriteStore;

    constructor() {
        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);
        this.favorites = new FavoriteStore(this);
    }
};
