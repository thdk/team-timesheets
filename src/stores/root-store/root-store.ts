import { RouterStore } from "mobx-router";

import { IRegistrationsStore, RegistrationStore } from "../registration-store/registration-store";
import { IConfigStore, ConfigStore } from "../config-store";
import { IUserStore, UserStore } from "../user-store";
import { IViewStore, ViewStore } from "../view-store";
import { IReportStore, ReportStore } from "../report-store/report-store";
import { DashboardStore, IDashboardStore } from "../dashboard-store";
import { ProjectStore } from "../project-store";
import { FavoriteStore } from "../favorite-store";
import { DivisionStore  } from "../division-store/division-store";

export interface IRootStore extends Store { };

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router: RouterStore<IRootStore>;
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: ProjectStore;
    public readonly favorites: FavoriteStore;
    public readonly divisions: DivisionStore;

    constructor({
        auth,
        firestore,
        httpsCallable,
    }: {
        firestore: firebase.firestore.Firestore,
        httpsCallable?: (name: string) => firebase.functions.HttpsCallable,
        auth?: firebase.auth.Auth,
    }) {
        this.user = new UserStore(
            this,
            {
                auth,
                firestore,
            }
        );

        this.view = new ViewStore();
        this.config = new ConfigStore(
            this,
            {
                firestore,
            },
        );
        this.timesheets = new RegistrationStore(
            this,
            {
                firestore,
            },
        );
        this.reports = new ReportStore(
            this,
            {
                firestore,
            },
        );
        this.dashboard = new DashboardStore(
            this,
            {
                firestore,
            },
        );
        this.projects = new ProjectStore(
            this,
            {
                firestore,
            },
        );
        this.favorites = new FavoriteStore(
            this,
            {
                firestore,
            },
        );
        this.divisions = new DivisionStore(
            this,
            {
                firestore,
                httpsCallable,
            },
        );
        this.router = new RouterStore<IRootStore>(this);
    }

    public dispose() {
        this.config.dispose();
        // this.dashboard.dispose();
        // this.favorites.dispose();
        // this.projects.dispose();
        // this.reports.dispose();
        this.timesheets.dispose();
        this.user.dispose();
        // this.view.dispose();
    }
};
