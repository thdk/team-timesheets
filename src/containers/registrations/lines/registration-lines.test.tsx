import React from "react";

import { Store } from "../../../stores/root-store";
import { initTestFirestore, deleteFirebaseAppsAsync } from "../../../__tests__/utils/firebase";
import { RegistrationLines } from "./registration-lines";
import { render } from "@testing-library/react";
import { Doc } from "firestorable";
import { IRegistration } from "../../../../common";
import userEvent from "@testing-library/user-event";

const {
    firestore,
    clearFirestoreDataAsync,
} = initTestFirestore("registrations-line-test");

const store = new Store({
    firestore,
});

jest.mock("../../../contexts/store-context", () => ({
    useStore: () => store,
}));

let clientIds: string[];
let taskIds: string[];
let projectIds: string[];

const setupAsync = async () => {
    taskIds = await store.config.tasksCollection.addAsync([
        {
            name: "Task 1",
            icon: "people",
        },
        {
            name: "Task 2",
        },
    ]);
    clientIds = await store.config.clientsCollection.addAsync([
        {
            name: "Client 1",
        },
    ]);

    projectIds = await store.projects.projectsCollection.addAsync([
        {
            name: "Project 1",
            icon: "people",
        },
    ]);
};

beforeAll(async () => {
    await clearFirestoreDataAsync();
    await setupAsync();
});
afterAll(() => {
    store.dispose();
    return Promise.all([
        deleteFirebaseAppsAsync(),
    ])
});

describe("RegistrationLines", () => {
    const getRegistrations = () => [
        {
            data: {
                userId: "user-1",
                client: clientIds[0],
                project: projectIds[0],
                task: taskIds[0],
                description: "Registration 1",
                date: new Date(2020, 2, 22),
            },
            id: "reg-1",
        },
        {
            data: {
                userId: "user-1",
                client: clientIds[0],
                project: projectIds[0],
                task: taskIds[2],
                description: "Registration 2",
                date: new Date(2020, 2, 22),
                time: 7.5
            },
            id: "reg-2",
        },
        {
            data: {},
            id: "reg-3",
            time: 7.555
        },
        {
            id: "reg-4",
        },
    ] as Doc<IRegistration>[];

    it("should render without registrations", () => {
        const { asFragment } = render(
            <RegistrationLines
                registrations={[]}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render registrations", () => {
        const { asFragment } = render(
            <RegistrationLines
                registrations={getRegistrations()}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should show checkbox for each registration when registrationToggleSelect is provided", () => {
        const registrationToggleSelect = jest.fn();
        const { container } = render(
            <RegistrationLines
                registrations={getRegistrations()}
                registrationToggleSelect={registrationToggleSelect}
            />
        );

        expect(
            container.querySelectorAll("input[type=checkbox]").length
        ).toBe(3);
    });

    it("should call callbacks for registration click and select", () => {
        const registrationToggleSelect = jest.fn();
        const registrationClick = jest.fn();

        const { container, getByText } = render(
            <RegistrationLines
                registrations={getRegistrations()}
                registrationToggleSelect={registrationToggleSelect}
                registrationClick={registrationClick}
            />
        );

        const checkboxEl = container.querySelectorAll("input[type=checkbox]")[0];

        userEvent.click(checkboxEl);

        expect(registrationToggleSelect).toBeCalledWith("reg-1");

        userEvent.click(getByText("Client 1 - Registration 2"));

        expect(registrationClick).toBeCalledWith("reg-2");
    });
});