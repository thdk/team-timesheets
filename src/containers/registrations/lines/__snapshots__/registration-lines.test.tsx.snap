// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationLines should render registrations 1`] = `
<DocumentFragment>
  <div
    class="data-row registration-line registration-line--readonly"
  >
    <div
      class="data-row__header"
    >
      <i
        class="rmwc-icon rmwc-icon--ligature material-icons"
        title="Task 1"
      >
        people
      </i>
    </div>
    <div
      class="data-row__column registration-line__time"
    >
      0.00
    </div>
    <div
      class="data-row__column registration-line__content"
    >
      <span
        class="data-row__line1 mdc-typography--subtitle1"
      >
        Project 1
      </span>
      <span
        class="data-row__line2 mdc-typography--subtitle2"
      >
        Client 1 - Registration 1
      </span>
    </div>
  </div>
  <div
    class="data-row registration-line registration-line--readonly"
  >
    <div
      class="data-row__column registration-line__time"
    >
      7.50
    </div>
    <div
      class="data-row__column registration-line__content"
    >
      <span
        class="data-row__line1 mdc-typography--subtitle1"
      >
        Project 1
      </span>
      <span
        class="data-row__line2 mdc-typography--subtitle2"
      >
        Client 1 - Registration 2
      </span>
    </div>
  </div>
  <div
    class="data-row registration-line registration-line--readonly"
  >
    <div
      class="data-row__column registration-line__time"
    >
      0.00
    </div>
    <div
      class="data-row__column registration-line__content"
    >
      <span
        class="data-row__line2 mdc-typography--subtitle2"
      >
        ...
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RegistrationLines should render without registrations 1`] = `<DocumentFragment />`;
